packages/google_sign_in/lib/google_sign_in.dart

```diff
@@ -126,31 +124,63 @@ class GoogleSignIn {
     await _initialization;
     Map<String, dynamic> response = await _channel.invokeMethod(method);
     _currentUser = (response != null && response.isNotEmpty)
-        ? new GoogleSignInAccount._(this, response) : null;
+        ? new GoogleSignInAccount._(this, response)
+        : null;
     _streamController.add(_currentUser);
     return _currentUser;
   }
 
-  /// The currently signed in account, or null if the user is signed out
-  GoogleSignInAccount _currentUser;
+  /// Keeps track of the most recently scheduled method call.
+  Completer _lastMethodCompleter;
+  String _lastMethodName;
+
+  /// Adds call to [method] in a queue for execution.
+  ///
+  /// At most one in flight call is allowed to prevent concurrent (out of order)
+  /// updates to [currentUser] and [onCurrentUserChanged].
+  Future<GoogleSignInAccount> _addMethodCall(String method) {
+    if (_lastMethodCompleter == null) {
+      _lastMethodName = method;
+      _lastMethodCompleter = new Completer<GoogleSignInAccount>();
+      _callMethod(method).then(_lastMethodCompleter.complete,
+          onError: _lastMethodCompleter.completeError);
+      return _lastMethodCompleter.future;
+    }
+
+    // Return the same Future for consecutive calls to the same method.
+    if (_lastMethodName == method && !_lastMethodCompleter.isCompleted) {
+      return _lastMethodCompleter.future;
+    }
+
+    var completer = new Completer<GoogleSignInAccount>();
+    _lastMethodCompleter.future.whenComplete(() {
+      _callMethod(method)
+          .then(completer.complete, onError: completer.completeError);
+    }).catchError((_) {
+      // Ignore if previous call completed with an error.
+    });
+    _lastMethodCompleter = completer;
+    _lastMethodName = method;
+    return _lastMethodCompleter.future;
+  }
+
+  /// The currently signed in account, or null if the user is signed out.
   GoogleSignInAccount get currentUser => _currentUser;
+  GoogleSignInAccount _currentUser;
 
   /// Attempts to sign in a previously authenticated user without interaction.
-  ///
-  /// If there is already a signed-in user, returns it.
-  Future<GoogleSignInAccount> signInSilently() async => _currentUser ?? _callMethod('signInSilently');
+  Future<GoogleSignInAccount> signInSilently() =>
+      _addMethodCall('signInSilently');
 
   /// Starts the sign-in process.
-  ///
-  /// If there is already a signed-in user, returns it.
-  Future<GoogleSignInAccount> signIn() async => _currentUser ?? _callMethod('signIn');
+  Future<GoogleSignInAccount> signIn() => _addMethodCall('signIn');
```

Comment:
For my own understanding: This is changing the logic slightly, right? Previously, if you called `signInSilently`, waited for the user object, and then for whatever reason called `signIn` you got the same user object back. With this change, the call to `signIn` will actually be dispatched to the native side (will it bring up a new sign-in dialog there?). Is this change intended?